### PR TITLE
Logger refactor: pass down as a param per factory instance

### PR DESCRIPTION
**Pull request by EmilianoSanchez**
# Javascript commons library

Changes:
- logger is passed down to modules:
  - via settings.log reference
  - and as a new param on modules that doesn't receive the settings object. For consistency, it is passed as the first param. 
- updated InLocalStorage module to use the injected logger
- removed `displayAllErrors` option. It was not being used (overwritten by default options), and anyway we can force logging them with some param in logger methods themselves
- removed `inputValidationLog` own log in segmentsSyncTask, since it now receives the logger as a param.
- removed `logger` own log in TimeTracker, since it now receives the logger as a param.
- removed `ownLog` in /logger/sdkLogger, since LoggerAPI can now use the injected logger instead of `ownLog`.
- removed `logFactory` and `defaultOptions` in /logger/sdkLogger, because it duplicates the logic of `Logger` and `defaultOptions` in /logger/index

### CODE REVIEW

File: src/storages/inLocalStorage/index.ts [hunk 1]
```diff
@@ -21,22 +22,23 @@ export interface InLocalStorageOptions {
...
-  if (!isLocalStorageAvailable()) {
-    log.warn('LocalStorage API is unavailable. Fallbacking into default MEMORY storage');
-    return;
-  }
-
   const prefix = options.prefix ? options.prefix + '.SPLITIO' : 'SPLITIO';
 
   return function InLocalStorageCSFactory(params: IStorageFactoryParams): IStorageSyncCS {
 
+    // Fallback to InMemoryStorage if LocalStorage API is not available
```
Comment (EmilianoSanchez) [inline src/storages/inLocalStorage/index.ts:29]: @NicoZelaya  take a look

File: src/evaluator/matchers/all.ts [hunk 2]
```diff
@@ -1,12 +1,11 @@
-import { logFactory } from '../../logger/sdkLogger';
-const log = logFactory('splitio-engine:matcher');
+// import { logFactory } from '../../logger/sdkLogger';
```
Comment (NicoZelaya) [inline src/evaluator/matchers/all.ts:1]: **General comment for matchers**

Let's analyze this ones when polishing or trimming messages, but we might be good logging from the matcherFactory only the type of matcher and the result. 
We loose visibility though (on the specific attributes used for the matching), so might be good to keep until we can have a more verbose build without affecting the main one. Or we can just leave a code comment and get back to it when we work on the verbose/debug build

File: src/evaluator/matchers/dependency.ts [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
+    log.debug(`[dependencyMatcher] Parent split "${parentName}" evaluated to "${evaluation.treatment}" with label "${evaluation.label}". ${parentName} evaluated treatment is part of [${acceptableTreatments}] ? ${matches}.`);
 
-export default function dependencyMatcherContext({ split, treatments }: IDependencyMatcherData, storage: IStorageSync | IStorageAsync) {
+    return matches;
+  }
 
   return function dependencyMatcher({ key, attributes }: IDependencyMatcherValue, splitEvaluator: ISplitEvaluator): MaybeThenable<boolean> {
     log.debug(`[dependencyMatcher] will evaluate parent split: "${split}" with key: ${JSON.stringify(key)} ${attributes ? `\n attributes: ${JSON.stringify(attributes)}` : ''}`);
-    const evaluation = splitEvaluator(key, split, attributes, storage);
+    const evaluation = splitEvaluator(log, key, split, attributes, storage);
```
Comment (NicoZelaya) [inline src/evaluator/matchers/dependency.ts:25]: This is the one we need to double check we're solving for, if we remove the logs from the matchers.

File: src/utils/settingsValidation/index.ts [hunk 4]
```diff
@@ -108,6 +111,10 @@ export function settingsValidation(config: unknown, validationParams: ISettingsV
   // creates a settings object merging base, defaults and config objects.
   const withDefaults = merge({}, base, defaults, config) as ISettingsInternal;
 
+  // ensure a valid logger
+  const log = logFactory('splitio'); // @ts-ignore
```
Comment (NicoZelaya) [inline src/utils/settingsValidation/index.ts:115]: Maybe we can try to add a comment with the reason each time we need a @ts-ignore

File: src/utils/timeTracker/index.ts [hunk 5]
```diff
@@ -196,16 +193,17 @@ const TrackerAPI = {
...
    */
-  stop(task: string, modifier?: number | string) {
+  stop(log: ILogger, task: string, modifier?: number | string) {
     const timerName = generateTimerKey(task, modifier);
     const timerData = timers[timerName];
     if (timerData) {
       // Stop the timer and round result for readability.
       const et = timerData.timer();
-      logger.debug(`[${task}] took ${et}ms to finish.`);
+      log.debug(`time tracker: [${task}] took ${et}ms to finish.`);
```
Comment (NicoZelaya) [inline src/utils/timeTracker/index.ts:206]: Use the same format than before. `[TIME TRACKER]`

**Review by NicoZelaya — APPROVED** [review]
Looks good to me, but let's not forget to try and reduce the param propagation in the upcoming PRs.

File: src/evaluator/combiners/and.ts [hunk 6]
```diff
@@ -1,19 +1,20 @@
...
 
-  log.debug(`[andCombiner] evaluates to ${hasMatchedAll}`);
-  return hasMatchedAll;
-}
+  function andResults(results: boolean[]): boolean {
+    // Array.prototype.every is supported by target environments
+    const hasMatchedAll = results.every(value => value);
 
-export default function andCombinerContext(matchers: IMatcher[]) {
+    log.debug(`[andCombiner] evaluates to ${hasMatchedAll}`);
```
[the remark this answers is not in the document]
Comment (EmilianoSanchez) [reply src/evaluator/combiners/and.ts:15]: thanks nico, I will review these suggestions after checking log messages in coming PR.

File: src/storages/AbstractSplitsCacheSync.ts [hunk 7]
```diff
@@ -99,7 +99,8 @@ export default abstract class AbstractSplitsCacheSync implements ISplitsCacheSyn
  * Given a parsed split, it returns a boolean flagging if its conditions use segments matchers (rules & whitelists).
  * This util is intended to simplify the implementation of `splitsCache::usesSegments` method
  */
-export function usesSegments({ conditions = [] }: ISplit) {
+export function usesSegments(split: ISplit) {
+  const conditions = split.conditions || [];
```
[the remark this answers is not in the document]
Comment (EmilianoSanchez) [reply src/storages/AbstractSplitsCacheSync.ts:103]: this change was not necessary (I can roll it back). 
But what you mentioned is right: both fetched and parsed offline splits always contains an array as "conditions" property. But `usesSegments` was done this way to simplify UTs (no need to define the conditions property in mock split definitions).

**Pull request merged**